### PR TITLE
golang: remove dependencies on go-ethereum

### PR DESCRIPTION
**Pull request by mandrigin**
This PR introduces `Hash` and `Address` types for Golang binding and removed dependencies on `go-ethereum/common` package.

### CODE REVIEW

File: bindings/go/evmc/types.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+// Lengths of hashes and addresses in bytes.
+const (
+	// hashLength is the expected length of the hash (in bytes)
+	hashLength = 32
+	// addressLength is the expected length of the address (in bytes)
+	addressLength = 20
+)
+
+// Hash represents the 32 byte Keccak256 hash of arbitrary data.
```
Comment (chfast) [inline bindings/go/evmc/types.go:13]:
```suggestion
// Hash represents the 32 bytes of arbitrary data (e.g. the result of Keccak256 hash).
```

File: bindings/go/evmc/host.go [hunk 2]
```diff
@@ -1,4 +1,3 @@
-// EVMC: Ethereum Client-VM Connector API.
```
Comment (chfast) [inline bindings/go/evmc/host.go:1]: Why?

File: bindings/go/evmc/host.go [hunk 3]
```diff
@@ -1,4 +1,3 @@
...
 // Copyright 2018-2019 The EVMC Authors.
```
Comment (chfast) [inline bindings/go/evmc/host.go:1]: Please bump the year to 2020.

File: bindings/go/evmc/types.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package evmc
```
Comment (chfast) [inline bindings/go/evmc/types.go:1]:
```suggestion
// EVMC: Ethereum Client-VM Connector API.
// Copyright 2020 The EVMC Authors.
// Licensed under the Apache License, Version 2.0.
package evmc
```

File: bindings/go/evmc/host_test.go [hunk 5]
```diff
@@ -51,18 +49,18 @@ func (host *testHostContext) GetTxContext() TxContext {
...
 }
 
-func (host *testHostContext) EmitLog(addr common.Address, topics []common.Hash, data []byte) {
+func (host *testHostContext) EmitLog(addr Address, topics []Hash, data []byte) {
 }
 
 func (host *testHostContext) Call(kind CallKind,
-	destination common.Address, sender common.Address, value *big.Int, input []byte, gas int64, depth int,
-	static bool, salt *big.Int) (output []byte, gasLeft int64, createAddr common.Address, err error) {
+	destination Address, sender Address, value *big.Int, input []byte, gas int64, depth int,
```
Comment (chfast) [inline bindings/go/evmc/host_test.go:60]: As we are changing types anyway, we may change `value *big.Int` to `value Hash` or better to `value Uint256BE` where `type Uint256BE = Hash`.

File: bindings/go/evmc/host_test.go [hunk 6]
```diff
@@ -51,18 +49,18 @@ func (host *testHostContext) GetTxContext() TxContext {
...
+	static bool, salt *big.Int) (output []byte, gasLeft int64, createAddr Address, err error) {
```
Comment (chfast) [inline bindings/go/evmc/host_test.go:61]: We definitely want to change `salt *big.Int` to `salt Hash`.

File: bindings/go/evmc/host.go (hunk 2 above)
Comment (mandrigin) [reply bindings/go/evmc/host.go:1]: I probably commited it by mistake

Comment (mandrigin) [reply bindings/go/evmc/host.go:1]: put the comment back

**Comment by mandrigin** [comment]
@chfast I think I addressed your feedback: updated the comments and got rid of `big.Int`

File: bindings/go/evmc/types.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+	hashLength = 32
+	// addressLength is the expected length of the address (in bytes)
+	addressLength = 20
+)
+
+// Hash represents the 32 bytes of arbitrary data (e.g. the result of Keccak256 hash).
+type Hash [hashLength]byte
+
+// Address represents the 20 byte address of an Ethereum account.
+type Address [addressLength]byte
```
Comment (axic) [inline bindings/go/evmc/types.go:18]: Rookie go question: are the const needed here or could something like `Address.len` work?

Comment (chfast) [reply bindings/go/evmc/types.go:18]: I think `len(Address)` should work in `const` context in asserts...

Comment (mandrigin) [reply bindings/go/evmc/types.go:18]: I’m not sure I understand the question. 
You mean just do `type Address [20]byte` here and use `len(Address)` in the other places? It should work theoretically. Here I mostly repeated how Go Ethereum does it.

**Comment by mandrigin** [comment]
@chfast done

File: bindings/go/evmc/types.go (hunk 7 above)
Comment (axic) [reply bindings/go/evmc/types.go:18]: > You mean just do type Address [20]byte here and use len(Address) in the other places?

Yes I meant that. It seems nicer from my side, but I have no strong preference.

Comment (mandrigin) [reply bindings/go/evmc/types.go:18]: I kept it as close as in go-ethereum as possible, and I don't have any strong preferences there either.

Comment (axic) [reply bindings/go/evmc/types.go:18]: @chfast you decide.

**Comment by mandrigin** [comment]
@chfast is there anything else I need to do here?

**Comment by chfast** [comment]
> @chfast is there anything else I need to do here?

No. I will be back on this soon. I want to try this `addressLen` change and will let you see it and comment on when ready.

**Comment by chfast** [comment]
@mandrigin Can you enable "Allow edits from maintainers" in this PR?

**Comment by mandrigin** [comment]
@chfast hmm, how do I do that after the fact? or should I reopen the PR?

**Comment by chfast** [comment]
> @chfast hmm, how do I do that after the fact? or should I reopen the PR?

@mandrigin There should be an option to that in the right column?

**Comment by mandrigin** [comment]
@chfast I don't see this option. Maybe it is because I'm trying to make a PR from a fork?
but if you want, I can try removing `addressLen` myself, it isn't that hard.

image

**Comment by chfast** [comment]
Ok, I pushed it here: https://github.com/ethereum/evmc/pull/513.

**Comment by mandrigin** [comment]
closing this one because it was superseded by #513

**Pull request closed without merging**